graphql-dgs-client/src/main/kotlin/com/netflix/graphql/dgs/client/WebsocketGraphQLClient.kt

```diff
@@ -0,0 +1,280 @@
+     * @param message The OperationMessage to send
+     */
+    fun send(message: OperationMessage) {
+        outgoingSink
+            .tryEmitNext(message)
+            .orThrow()
+    }
+
+    /**
+     * Stream messages from the server, lazily establish connection
+     * @return Flux of OperationMessages
+     */
+    fun receive(): Flux<OperationMessage> {
+        return Flux.defer {
+            conn.subscribe()
+            incomingSink.asFlux()
+        }
+    }
+
+    private fun exchange(
+        incomingMessages: Sinks.Many<OperationMessage>,
+        outgoingMessages: Sinks.Many<OperationMessage>,
+        session: WebSocketSession): Mono<Void> {
+        // Create chains to handle de/serialization
+        val incomingDeserialized = session
+            .receive()
+            .map(this::decodeMessage)
+            .doOnNext(incomingMessages::tryEmitNext)
+        val outgoingSerialized = session
+            .send(outgoingMessages
+                .asFlux()
+                .map{ createMessage(session, it) })
+
+        // Transfer the contents of the sinks to/from the server
+        return Flux
+            .merge(incomingDeserialized, outgoingSerialized)
+            .then()
+    }
+
+    private fun createMessage(
+        session: WebSocketSession,
+        message: OperationMessage): WebSocketMessage {
+
+        return session.textMessage(MAPPER.writeValueAsString(message))
+    }
+
+    private fun decodeMessage(message: WebSocketMessage): OperationMessage {
+        val messageText = message.payloadAsText;
+        val type = object : TypeReference<OperationMessage>() {}
+
+        return MAPPER.readValue(messageText, type)
+    }
+}
+
+
+const val GQL_CONNECTION_INIT = "connection_init"
+const val GQL_CONNECTION_ACK = "connection_ack"
+const val GQL_CONNECTION_ERROR = "connection_error"
+const val GQL_START = "start"
+const val GQL_STOP = "stop"
+const val GQL_DATA = "data"
+const val GQL_ERROR = "error"
+const val GQL_COMPLETE = "complete"
+const val GQL_CONNECTION_TERMINATE = "connection_terminate"
+const val GQL_CONNECTION_KEEP_ALIVE = "ka"
+
+data class OperationMessage(
+    @JsonProperty("type")
+    val type: String,
+    @JsonProperty("payload")
+    @JsonTypeInfo(use = JsonTypeInfo.Id.DEDUCTION)
+    @JsonSubTypes(
+        JsonSubTypes.Type(value = DataPayload::class),
+        JsonSubTypes.Type(value = QueryPayload::class))
+    val payload: Any? = null,
+    @JsonProperty("id", required = false)
+    val id: String? = "")
+
+data class DataPayload(
+    @JsonProperty("data")
+    val data: Any?,
+    @JsonProperty("errors")
+    val errors: List<Any>? = emptyList())
+
+data class QueryPayload(
+    @JsonProperty("variables")
+    val variables: Map<String, Any>?,
+    @JsonProperty("extensions")
+    val extensions: Map<String, Any> = emptyMap(),
+    @JsonProperty("operationName")
+    val operationName: String?,
+    @JsonProperty("query")
+    val query: String)
```

Comment:
@osi These are near-duplicates of the structures in `graphql-dgs-subscriptions-websockets`. Is there somewhere common they can live?

Reply:
I'll defer to @paulbakker  - may be an opportunity for a new shared module with common subscription stuff.